pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -20,148 +20,296 @@
+          // if after adding to the set, the number of predicate values become different (less),
+          // it implies that values got deduplicated and so we need to rewrite this predicate later
+          if (predicateValues.size() != childValues.size()) {
+            rewriteINEQChildrenOfORRootOperator = true;
           }
+          columnToInEqPredicateValues.put(column, predicateValues);
         } else {
-          Set<String> currentValues = columnToValues.get(childQueryTree.getColumn());
-          for (String childValue : childValues) {
-            if (!containsDuplicates && currentValues.contains(childValue)) {
-              containsDuplicates = true;
-            } else {
-              currentValues.add(childValue);
-            }
-          }
+          // if we had earlier seen this column, we need to rewrite the predicate
+          predicateValues.addAll(childValues);
+          rewriteINEQChildrenOfORRootOperator = true;
         }
-      } else {
-        nonEqualityOperators.add(childQueryTree);
-      }
-    }
-
-    return containsDuplicates;
-  }
 
-  private boolean isRebuildRequired(Map<String, Set<String>> columnToValues, boolean containsDuplicates) {
-    // We need to rebuild the predicate if there were duplicate values detected (eg. a = 1 OR a = 1) or if there is
-    // more than one value for a column (eg. a = 1 OR a = 2)
-    boolean rebuildRequired = containsDuplicates;
-
-    if (!rebuildRequired) {
-      for (Set<String> columnValues : columnToValues.values()) {
-        if (1 < columnValues.size()) {
-          rebuildRequired = true;
-          break;
+        if (!rewriteINEQChildrenOfORRootOperator && operator == FilterOperator.IN && predicateValues.size() == 1) {
+          // single predicate value IN should be rewritten
+          rewriteINEQChildrenOfORRootOperator = true;
+        }
+      } else {
+        // the immediate child of root operator is <, >, AND, <=, >=, BETWEEN ...
+        if (childQueryTree.getOperator() == FilterOperator.AND) {
+          nonEqInOperators.add(optimize(childQueryTree));
+        } else {
+          nonEqInOperators.add(childQueryTree);
         }
       }
     }
-    return rebuildRequired;
+    return rewriteINEQChildrenOfORRootOperator;
   }
 
-  private FilterQueryTree rebuildFilterPredicate(Map<String, Set<String>> columnToValues,
-      List<FilterQueryTree> nonEqualityOperators) {
+  private FilterQueryTree rebuildFilterPredicate(Map<String, Set<String>> columnToInEqPredicateValues) {
     ArrayList<FilterQueryTree> newChildren = new ArrayList<>();
-    for (Map.Entry<String, Set<String>> columnAndValues : columnToValues.entrySet()) {
-      newChildren.add(buildFilterQueryTreeForColumnAndValues(columnAndValues));
+    for (Map.Entry<String, Set<String>> columnAndPredicate : columnToInEqPredicateValues.entrySet()) {
+      newChildren.add(buildFilterQueryTreeForColumnAndPredicateInfo(columnAndPredicate.getKey(), columnAndPredicate.getValue()));
```

Comment:
Inline this

Reply:
This is reusable -- called for the when there are multiple predicates on single column only and we can remove OR root all-together. So buildFilterQueryTreeForColumnAndPredicateInfo helps by creating the new tree with IN/EQ operator.

Reply:
You can inline rebuildFilterPredicate() and pre-allocate the size of newChildren. Your choice